src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,797 @@
-      Tensor<xpu, 1, DType> ograd = inputs[0].data().FlatTo1D<xpu, DType>(s);
-      ASSIGN_DISPATCH(lgrad, req[0], F<LOP>(ograd));
-      ASSIGN_DISPATCH(rgrad, req[1], F<ROP>(ograd));
-      ASSIGN_DISPATCH(lgrad_idx, req[0], F<LOP>(ograd_idx));
-      ASSIGN_DISPATCH(rgrad_idx, req[1], F<ROP>(ograd_idx));
+
+ public:
+  template<typename xpu, typename OP>
+  static void Launch(const nnvm::NodeAttrs &attrs,
+                     const OpContext &ctx,
+                     const std::vector<TBlob> &inputs,
+                     const std::vector<OpReqType> &req,
+                     const std::vector<TBlob> &outputs) {
+    using namespace mxnet_op;
+    if (req[0] != kNullOp) {
+      Stream<xpu> *s = ctx.get_stream<xpu>();
+      CHECK_EQ(inputs.size(), 2U);
+      CHECK_EQ(outputs.size(), 1U);
+      MXNET_ASSIGN_REQ_SWITCH(req[0], Req, {
+        MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+          const size_t size = (minthree(outputs[0].Size(), inputs[0].Size(), inputs[1].Size())
+          + DataType<DType>::kLanes - 1) / DataType<DType>::kLanes;
+          Kernel<BMap<OP, Req>, xpu>::Launch(s, size,
+          outputs[0].dptr<DType>(),
+          inputs[0].dptr<DType>(), inputs[1].dptr<DType>());
+        });
+      });
+    }
+  }
+
+  template<typename xpu, typename OP>
+  static void LaunchWithHalf2(const nnvm::NodeAttrs &attrs,
+                     const OpContext &ctx,
+                     const std::vector<TBlob> &inputs,
+                     const std::vector<OpReqType> &req,
+                     const std::vector<TBlob> &outputs) {
+    using namespace mxnet_op;
+    if (req[0] != kNullOp) {
+      Stream<xpu> *s = ctx.get_stream<xpu>();
+      CHECK_EQ(inputs.size(), 2U);
+      CHECK_EQ(outputs.size(), 1U);
+      MXNET_ASSIGN_REQ_SWITCH(req[0], Req, {
+        MSHADOW_TYPE_SWITCH_WITH_HALF2(outputs[0].type_flag_, DType, {
+          const size_t size = (minthree(outputs[0].Size(), inputs[0].Size(), inputs[1].Size())
+          + DataType<DType>::kLanes - 1) / DataType<DType>::kLanes;
+          Kernel<BMap<OP, Req>, xpu>::Launch(s, size,
+          outputs[0].dptr<DType>(),
+          inputs[0].dptr<DType>(), inputs[1].dptr<DType>());
+        });
+      });
+    }
+  }
+
+  template<typename xpu, typename OP>
+  static void LaunchEx(const nnvm::NodeAttrs &attrs,
+                       const OpContext &ctx,
+                       const std::vector<NDArray> &inputs,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<NDArray> &outputs) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(inputs.size(), 2);
+    CHECK_EQ(outputs.size(), 1);
+    if (req[0] != kNullOp) {
+      // If any input or output is dense, fallback to FCompute
+      if (!common::ContainsDefaultStorage(inputs)) {
+        mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+        switch (inputs[0].storage_type()) {
+          case kRowSparseStorage:
```

Comment:
I think Haibin was referring to `storage_type`, instead of `dtype`. If that's the case, could you add a CECK_EQ for `inputs[0]` and `inputs[1]` storage types?